### PR TITLE
d1: add new target

### PR DESCRIPTION
**Pull request by wigyori**
This target adds support for the Allwinner D1 RISC-V based SoCs, based on 6.1 and patches backported from later kernels.

 - RISC-V single-core T-Head C906 (RV64GCV)
 - Tensilica HiFi4 DSP
 - DDR2/DDR3 support
 - 10/100/1000M ethernet
 - usual peripherals like USB2, SPI, I2C, PWM, etc.

Four boards are supported:
 - Dongshan Nezha STU
    - 512Mb RAM
    - ethernet

 - LicheePi RV Dock
    - 512Mb RAM
    - wireless-only (RTL8723DS)

 - MangoPi MQ-Pro
    - 512Mb RAM
    - there are pads available for an SPI flash
    - wireless-only (RTL8723DS)

 - Nezha D1
    - 512Mb/1Gb/2Gb RAM
    - 256Mb NAND flash
    - ethernet, wireless

Installation:
Standard SD-card installation via dd-ing the generated image to
an SD-card of at least 256Mb.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

### CODE REVIEW

**Comment by kassane** [comment]
**Note**: Nezha use xr289 (xradio) - proprietary code.

https://github.com/tmolteno/d1_build/issues/7

**Comment by wigyori** [comment]
Rebased target, upgraded u-boot to 2023.01, cleaned up kernel config.

**Comment by wigyori** [comment]
Any takers for reviewing?

**Comment by wigyori** [comment]
This PR is submitted for a long time now, being rebased every now and then. As there were no change requests or comments, I take it as the target is looking good. Unless there are further comments, I will merge this target on 16/Feb.

**Comment by aiamadeus** [comment]
Thank you very much for your work, having d1 as the target name is a bit strange.
Allwinner company may have RISC-V SoCs with other names in the future, so I think
it would be better to change the target name?

**Comment by wigyori** [comment]
> Thank you very much for your work, having d1 as the target name is a bit strange. Allwinner company may have RISC-V SoCs with other names in the future, so I think it would be better to change the target name?

I'd say targets can be renamed if needed, just like it was with the imx23->mxs and the imx6->imx change. Right now I'm not aware if Allwinner has future RISC-V SoCs in the pipeline, if they come up with one, we'll see about the naming - the only roadmap I was able to find had ARMs left, right and centre.

File: package/boot/uboot-d1/Makefile [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#
+
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+
+PKG_VERSION:=2023.01
+PKG_RELEASE:=1
+
+#PKG_HASH:=352ad311dcf9431fc112122b0afe710e3104c0e670fc943273e8ed3b65c17335
+PKG_HASH:=69423bad380f89a0916636e89e6dcbd2e4512d584308d922d1039d1e4331950f
```
Comment (linusw) [inline package/boot/uboot-d1/Makefile:13]: Why a commented out hash?

**Review by linusw — APPROVED** [review]
Just a minor comment otherwise looks fine. I assume all the four-digit patches are from some branch you will try to at least partially push to U-Boot so it will get more review at that point.

File: package/boot/uboot-d1/Makefile (hunk 1 above)
Comment (wigyori) [reply package/boot/uboot-d1/Makefile:13]: Leftover from a previous u-boot version test - cleaned up, thanks.

**Comment by wigyori** [comment]
Merged, thanks all for checking, commenting and approving.

**Comment by robberphex** [comment]
I compile OpenWrt for LicheePi RV, but I cannot find `rtl8723ds` module in image, and I cannot choose it in `make menuconfig`.

Is there anyway to compile with rtl8723ds module?

**Comment by wigyori** [comment]
The rtl8723ds kmod could be easily added (it seems to be just a new package def in mac80211), but the firmware for the 8723ds is not uploaded to the linux-firmware repository yet. Before adding a new kmod, I'd rather like to see its firmware there - the fw files are available elsewhere on the 'net, but based on what I've found around it, there might be licensing questions around it. One suggestion I've found was to use the 8723D firmware - let me double-check that if that'll work.

**Comment by wigyori** [comment]
@robberphex - this staging branch [1] will get you going both for the wifi and the BT part.

The wifi part seems to be OK with the 8723D firmware, the missing rtl_bt/rtl8723ds_* are only a concern for the BT part. The latter will need to be added into the linux-firmware repo by Realtek before I can pick them up for adding the updated BT driver into our trunk. Based on that other similar Realtek firmwares are licensed as "Redistributable", I have added them as a hack into my staging.

[1] https://git.openwrt.org/?p=openwrt/staging/wigyori.git;a=shortlog;h=refs/heads/d1-bt

**Comment by robberphex** [comment]
Looks like rtl8723ds is builtin, in [d1-bt branch](https://git.openwrt.org/?p=openwrt/staging/wigyori.git;a=shortlog;h=refs/heads/d1-bt)?
98a3a7d6ddf4fc1eb6d3bb80720f894

I tried built image and boot in, still doesn't have network card. And it looks like no module `rtl8723ds`, but `rtw88_8723ds`, and network card doesn't work.

<details>
    <summary>Here are logs</summary>

```
root@OpenWrt:/# ifconfig -a
lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:384 errors:0 dropped:0 overruns:0 frame:0
          TX packets:384 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:29952 (29.2 KiB)  TX bytes:29952 (29.2 KiB)

root@OpenWrt:/# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
root@OpenWrt:/# lsmod
cfg80211              524288  2 rtw88_core,mac80211
cmac                   16384  0
compat                 16384  3 rtw88_8723ds,mac80211,cfg80211
crc_ccitt              16384  1 ppp_async
drbg                   32768  0
hmac                   16384  0
jitterentropy_rng      16384  0
libcrc32c              16384  1 nf_tables
mac80211             1077248  2 rtw88_sdio,rtw88_core
nf_conntrack          131072  7 nft_redir,nft_nat,nft_masq,nft_flow_offload,nft_ct,nf_nat,nf_flow_table
nf_defrag_ipv4         16384  1 nf_conntrack
nf_defrag_ipv6         24576  1 nf_conntrack
nf_flow_table          49152  2 nf_flow_table_inet,nft_flow_offload
nf_flow_table_inet     16384  0
nf_log_syslog          28672  0
nf_nat                 49152  4 nft_redir,nft_nat,nft_masq,nft_chain_nat
nf_reject_ipv4         16384  2 nft_reject_ipv4,nft_reject_inet
nf_reject_ipv6         16384  2 nft_reject_ipv6,nft_reject_inet
nf_socket_ipv4         16384  0
nf_socket_ipv6         16384  0
nf_tables             315392137 nft_fib_inet,nf_flow_table_inet,nft_reject_ipv6,nft_reject_ipv4,nft_reject_inet,nft_reject,nft_redir,nft_quota,nft_objref,nft_numgen,nft_nat,nft_masq,nft_log,nft_limit,nft_hash,nft_flow_offload,nft_fib_ipv6,nft_fib_ipv4,nft_fib,nft_ct,nft_chain_nat
nf_tproxy_ipv4         16384  0
nf_tproxy_ipv6         16384  0
nfnetlink              16384  1 nf_tables
nft_chain_nat          16384  2
nft_ct                 24576  3
nft_fib                16384  3 nft_fib_inet,nft_fib_ipv6,nft_fib_ipv4
nft_fib_inet           16384  0
nft_fib_ipv4           16384  1 nft_fib_inet
nft_fib_ipv6           16384  1 nft_fib_inet
nft_flow_offload       16384  0
nft_hash               16384  0
nft_limit              16384  5
nft_log                16384  0
nft_masq               16384  1
nft_nat                16384  0
nft_numgen             16384  0
nft_objref             16384  0
nft_quota              16384  0
nft_redir              16384  0
nft_reject             16384  3 nft_reject_ipv6,nft_reject_ipv4,nft_reject_inet
nft_reject_inet        16384  2
nft_reject_ipv4        16384  0
nft_reject_ipv6        16384  0
ppp_async              28672  0
ppp_generic            57344  3 pppoe,ppp_async,pppox
pppoe                  24576  0
pppox                  16384  1 pppoe
rtw88_8723d            45056  1 rtw88_8723ds
rtw88_8723ds           16384  0
rtw88_core            208896  2 rtw88_8723d,rtw88_sdio
rtw88_sdio             28672  1 rtw88_8723ds
seqiv                  16384  0
sha512_generic         16384  0
slhc                   16384  1 ppp_generic
tcp_bbr                24576  5
```
</details>

**Comment by robberphex** [comment]
<details>
    <summary>iw phy</summary>

```
root@OpenWrt:/# iw phy
Wiphy phy0
        wiphy index: 0
        max # scan SSIDs: 4
        max scan IEs length: 2257 bytes
        max # sched scan SSIDs: 0
        max # match sets: 0
        Retry short limit: 7
        Retry long limit: 4
        Coverage class: 0 (up to 0m)
        Device supports T-DLS.
        Available Antennas: TX 0x1 RX 0x1
        Configured Antennas: TX 0x1 RX 0x1
        Supported interface modes:
                 * IBSS
                 * managed
                 * AP
                 * AP/VLAN
                 * monitor
                 * mesh point
        Band 1:
                Capabilities: 0x196e
                        HT20/HT40
                        SM Power Save disabled
                        RX HT20 SGI
                        RX HT40 SGI
                        RX STBC 1-stream
                        Max AMSDU length: 7935 bytes
                        DSSS/CCK HT40
                Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
                Minimum RX AMPDU time spacing: 16 usec (0x07)
                HT Max RX data rate: 150 Mbps
                HT TX/RX MCS rate indexes supported: 0-7, 32
                Frequencies:
                        * 2412 MHz [1] (20.0 dBm)
                        * 2417 MHz [2] (20.0 dBm)
                        * 2422 MHz [3] (20.0 dBm)
                        * 2427 MHz [4] (20.0 dBm)
                        * 2432 MHz [5] (20.0 dBm)
                        * 2437 MHz [6] (20.0 dBm)
                        * 2442 MHz [7] (20.0 dBm)
                        * 2447 MHz [8] (20.0 dBm)
                        * 2452 MHz [9] (20.0 dBm)
                        * 2457 MHz [10] (20.0 dBm)
                        * 2462 MHz [11] (20.0 dBm)
                        * 2467 MHz [12] (20.0 dBm) (no IR)
                        * 2472 MHz [13] (20.0 dBm) (no IR)
                        * 2484 MHz [14] (20.0 dBm) (no IR)
        interface combinations are not supported
        HT Capability overrides:
                 * MCS: ff ff ff ff ff ff ff ff ff ff
                 * maximum A-MSDU length
                 * supported channel width
                 * short GI for 40 MHz
                 * max A-MPDU length exponent
                 * min MPDU start spacing
        max # scan plans: 1
        max scan plan interval: -1
        max scan plan iterations: 0
        Supported extended features:
                * [ RRM ]: RRM
                * [ SET_SCAN_DWELL ]: scan dwell setting
                * [ FILS_STA ]: STA FILS (Fast Initial Link Setup)
                * [ CONTROL_PORT_OVER_NL80211 ]: control port over nl80211
                * [ TXQS ]: FQ-CoDel-enabled intermediate TXQs
                * [ SCAN_RANDOM_SN ]: use random sequence numbers in scans
                * [ CAN_REPLACE_PTK0 ]: can safely replace PTK 0 when rekeying
                * [ CONTROL_PORT_NO_PREAUTH ]: disable pre-auth over nl80211 control port support
                * [ DEL_IBSS_STA ]: deletion of IBSS station support
                * [ SCAN_FREQ_KHZ ]: scan on kHz frequency support
                * [ CONTROL_PORT_OVER_NL80211_TX_STATUS ]: tx status for nl80211 control port support
```
</details>

**Comment by wigyori** [comment]
Thanks. Please remove the 'disable' flag in /etc/config/wireless, and try to bring up the wifi interface with "wifi up".

Note: various reports were seen that AP mode is buggy on this 8723DS, so it might or might not work.

**Comment by robberphex** [comment]
Sorry but didn't work.

```
root@OpenWrt:/# cat /etc/config/wireless

config wifi-device 'radio0'
        option type 'mac80211'
        option path 'platform/soc/4021000.mmc/mmc_host/mmc1/mmc1:0001/mmc1:0001:1'
        option channel '1'
        option band '2g'
        option htmode 'HT20'

config wifi-iface 'default_radio0'
        option device 'radio0'
        option network 'lan'
        option mode 'ap'
        option ssid 'OpenWrt'
        option encryption 'none'

root@OpenWrt:/# ifconfig -a
br-lan    Link encap:Ethernet  HWaddr 02:CD:E3:60:9E:39
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fd8a:512c:1a95::1/60 Scope:Global
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:128 errors:0 dropped:0 overruns:0 frame:0
          TX packets:128 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:9984 (9.7 KiB)  TX bytes:9984 (9.7 KiB)

root@OpenWrt:/# iw phy
Wiphy phy0
        wiphy index: 0
        max # scan SSIDs: 4
        max scan IEs length: 2257 bytes
        max # sched scan SSIDs: 0
        max # match sets: 0
        Retry short limit: 7
        Retry long limit: 4
        Coverage class: 0 (up to 0m)
        Device supports T-DLS.
        Available Antennas: TX 0x1 RX 0x1
        Configured Antennas: TX 0x1 RX 0x1
        Supported interface modes:
                 * IBSS
                 * managed
                 * AP
                 * AP/VLAN
                 * monitor
                 * mesh point
        Band 1:
                Capabilities: 0x196e
                        HT20/HT40
                        SM Power Save disabled
                        RX HT20 SGI
                        RX HT40 SGI
                        RX STBC 1-stream
                        Max AMSDU length: 7935 bytes
                        DSSS/CCK HT40
                Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
                Minimum RX AMPDU time spacing: 16 usec (0x07)
                HT Max RX data rate: 150 Mbps
                HT TX/RX MCS rate indexes supported: 0-7, 32
                Frequencies:
                        * 2412 MHz [1] (20.0 dBm)
                        * 2417 MHz [2] (20.0 dBm)
                        * 2422 MHz [3] (20.0 dBm)
                        * 2427 MHz [4] (20.0 dBm)
                        * 2432 MHz [5] (20.0 dBm)
                        * 2437 MHz [6] (20.0 dBm)
                        * 2442 MHz [7] (20.0 dBm)
                        * 2447 MHz [8] (20.0 dBm)
                        * 2452 MHz [9] (20.0 dBm)
                        * 2457 MHz [10] (20.0 dBm)
                        * 2462 MHz [11] (20.0 dBm)
                        * 2467 MHz [12] (20.0 dBm) (no IR)
                        * 2472 MHz [13] (20.0 dBm) (no IR)
                        * 2484 MHz [14] (20.0 dBm) (no IR)
        interface combinations are not supported
        HT Capability overrides:
                 * MCS: ff ff ff ff ff ff ff ff ff ff
                 * maximum A-MSDU length
                 * supported channel width
                 * short GI for 40 MHz
                 * max A-MPDU length exponent
                 * min MPDU start spacing
        max # scan plans: 1
        max scan plan interval: -1
        max scan plan iterations: 0
        Supported extended features:
                * [ RRM ]: RRM
                * [ SET_SCAN_DWELL ]: scan dwell setting
                * [ FILS_STA ]: STA FILS (Fast Initial Link Setup)
                * [ CONTROL_PORT_OVER_NL80211 ]: control port over nl80211
                * [ TXQS ]: FQ-CoDel-enabled intermediate TXQs
                * [ SCAN_RANDOM_SN ]: use random sequence numbers in scans
                * [ CAN_REPLACE_PTK0 ]: can safely replace PTK 0 when rekeying
                * [ CONTROL_PORT_NO_PREAUTH ]: disable pre-auth over nl80211 control port support
                * [ DEL_IBSS_STA ]: deletion of IBSS station support
                * [ SCAN_FREQ_KHZ ]: scan on kHz frequency support
                * [ CONTROL_PORT_OVER_NL80211_TX_STATUS ]: tx status for nl80211 control port support
root@OpenWrt:/#
```

**Pull request closed without merging**